src/components/PeriodDimension/FixedPeriodSelect.js

```diff
@@ -0,0 +1,89 @@
+import React, { useState, useEffect } from 'react'
+import propTypes from '@dhis2/prop-types'
+import i18n from '@dhis2/d2-i18n'
+import { SingleSelectField, SingleSelectOption } from '@dhis2/ui'
+import FixedPeriodFilter from './FixedPeriodFilter.js'
+import {
+    parsePeriodCode,
+    getFixedPeriodsOptionsById as getPeriodById,
+    getYearOffsetFromNow,
+} from './utils/fixedPeriods.js'
+import styles from './styles/FixedPeriodSelect.style.js'
+
+const FixedPeriodSelect = ({
+    onChange,
+    allowedPeriodTypes,
+    className,
+    dataTest,
+    value,
+}) => {
+    const [periodType, setPeriodType] = useState('')
+    const [year, setYear] = useState('')
+    const [options, setOptions] = useState()
+
+    useEffect(() => {
+        const period = parsePeriodCode(value, allowedPeriodTypes)
+        if (period) {
+            setPeriodType(period.id)
+            setYear(period.year)
+            setOptions(period.options)
+        } else if (periodType && year) {
+            const offset = getYearOffsetFromNow(year)
+            const options = getPeriodById(periodType).getPeriods({ offset })
+            setOptions(options)
+        }
+    }, [allowedPeriodTypes, periodType, value, year])
```

Comment:
I believe this will flash a single empty render before parsing and rendering the period, since `useEffect` happens after the first commit phase, is that correct?

Reply:
If so, is there a way to avoid this?

Reply:
I believe you're right. The first time `useEffect` will run is when the component mounts and this is the second render. Apart from changing this to a class component and also running this logic in the constructor, I wouldn't know any way of avoiding this.

I'm not sure if this is something we need to worry about though. It shouldn't be visible to the human eye, and I also think the current way is quite a common approach (i.e. either using `useEffect` or `componentDidMount` + `componentDidUpdate` (and not executing the same logic in the constructor).

Reply:
I've discussed this with @martinkrulltott and also looked into it a bit. When using CPU throttling x 6 I could visually observe the content flash. I guess since we're targeting low-end devices this is relevant. The initial solution looked especially bad, because the first render does produces a partial UI. Martin suggested that if it's not possible to completely avoid the flash, it would be better to just return null instead of a partial UI. I agreed with that....

So I looked into two solutions:
1. Return `null` instead of a partial UI. So we just don't show anything on the first render, and on the second render everything is as it should be. So this doesn't really solve the issue, but it does make it less bad. And on the plus side: we're still using a functional component and have not increased the line-count too much. See https://github.com/dhis2/analytics/pull/509/commits/ee5dd24e8c8c3f505c249cf33e3ce881354b7a67
2. Refactor to a class based component and use `getDerivedStateFromProps` to make sure the state is correct **before** the (first) render. It seems to me that this is actually working perfectly: I can't observe a content flash anymore, even with my CPU throttled x 6. The line-count has increased a quite a bit, but I would probably argue that the code is slightly more readable in its current form. See https://github.com/dhis2/analytics/pull/509/commits/66dd411131b8a273cc5045afa998c4a66bc5f2b6

I guess my personal preference would be to go with the class based component.

Reply:
@HendrikThePendric the state of this component is derived exclusively from the props, so I don't think this should use a `useEffect` at all and it should actually remain a simple, pure functional component.  Something like this:

```js
const FixedPeriodSelect = ({
    onChange,
    allowedPeriodTypes,
    className,
    dataTest,
    value,
}) => {
    const period = parsePeriodCode(value, allowedPeriodTypes) || {}
    const [periodType, setPeriodType] = useState(period.id)
    const [year, setYear] = useState(period.year)

    let options = period.options
    if (periodType && year) {
        const offset = getYearOffsetFromNow(year)
        options = getPeriodById(periodType).getPeriods({ offset })
    }
```

I'm not sure about the logic underlying the overwrite of `options`, but kept the existing behavior.  This only accepts the `value` and `allowedPeriodTypes` to be specified once on component mount, but I believe that's the intent anyway (otherwise the state will get overwritten).  It would be possible to add a `useEffect` for the update here as well, if needed.

Finally, this could be cleaned up quite a bit I think with a helper function (maybe as a singular state variable) but I'll leave that as a bonus here.

Reply:
If the intent is to allow this too be controlled or uncontrolled, then a `useEffect` with an `initialRender` state might work as well

Reply:
Here's an alternative approach:

```js
const parseOptions = period => {
    if (period.id && period.year) {
        const offset = getYearOffsetFromNow(year)
        return getPeriodById(periodType).getPeriods({ offset })
    }
    return period.options
}
const FixedPeriodSelect = ({
    onChange,
    allowedPeriodTypes,
    className,
    dataTest,
    value,
}) => {
    const [period, setPeriod] = useState(parsePeriodCode(value, allowedPeriodTypes))
    const options = parseOptions(period)

    // I would actually recommend AGAINST this, I don't think we should support both controlled and uncontrolled components...  not sure what the intent was here.  See https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html
    const useEffect(() => {
        const newPeriod = parsePeriodCode(value, allowedPeriodTypes)
        if (!shallowEqual(period, newPeriod)) {
            setPeriod(newPeriod)
        }
    }, [value, allowedPeriodTypes])
    
    // ..., use `setPeriod({ ...period, id: periodType })` etc. below
```

Reply:
>  the state of this component is derived exclusively from the props, so I don't think this should use a useEffect at all and it should actually remain a simple, pure functional component.

This is not entirely correct. There are two paths to consider:
1. There is a value: now the state period-type and year are fully derived
2. There is no value: now the state for period-type and year are not derived, and more importantly, user-interaction is required to set the period-type and year state. Only once these state properties both have a value, the actual period can be selected.

As such, I believe we do need state, and I think that the current implementation, which implements `getDerivedStateFromProps`, is a good fit.

Reply:
Nice, yeah the class Component version with `getDerivedStateFromProps` looks pretty clean to me

Reply:
🎉